core/unpack/unpacker.go

```diff
@@ -478,16 +576,41 @@ func (u *Unpacker) unpack(
 			tracing.Attribute("layer.media.size", desc.Size),
 			tracing.Attribute("layer.media.digest", desc.Digest.String()),
 		)
-		if err := doUnpackFn(i, desc); err != nil {
-			layerSpan.SetStatus(err)
+		statusCh, err := topHalf(i, desc, layerSpan, unpackLayerStart)
+		if err != nil {
+			if parallel {
+				break
+			} else {
+				layerSpan.SetStatus(err)
+				layerSpan.End()
+				return err
+			}
+		}
+		if statusCh == nil {
+			// nothing to do, already exists
 			layerSpan.End()
+			continue
+		}
+		if parallel {
+			statusChans = append(statusChans, statusCh)
+		} else {
+			if err = bottomHalf(<-statusCh, nil); err != nil {
+				return err
+			}
+		}
+	}
+
+	// In parallel mode, snapshots still need to be committed and rebased sequentially
+	if parallel {
+		var errs error
+		for _, sc := range statusChans {
+			if err := bottomHalf(<-sc, errs); err != nil {
+				errs = errors.Join(errs, err)
+			}
+		}
+		if errs != nil {
```

Comment:
Should this return `errs` and not `err`?

Reply:
Fixed